cognite/extractorutils/retry.py

```diff
@@ -0,0 +1,83 @@
+import logging
+import random
+import threading
+from functools import partial
+from typing import Iterable, Optional, Type
+
+from decorator import decorator
+
+logging_logger = logging.getLogger(__name__)
+
+
+def _retry_internal(
+    f,
+    cancelation_token: threading.Event = threading.Event(),
+    exceptions: Iterable[Type[Exception]] = Exception,
+    tries: int = -1,
+    delay: int = 0,
+    max_delay: Optional[int] = None,
+    backoff: float = 1,
+    jitter: float = 0,
+    logger: logging.Logger = logging_logger,
+):
+    _tries, _delay = tries, delay
+    while _tries and not cancelation_token.is_set():
+        try:
+            return f()
+        except exceptions as e:
+            _tries -= 1
+            if not _tries:
+                raise
+
+            if logger is not None:
+                logger.warning("%s, retrying in %s seconds...", e, _delay)
+
+            cancelation_token.wait(_delay)
+            _delay *= backoff
+
+            if isinstance(jitter, tuple):
+                _delay += random.uniform(*jitter)
+            else:
+                _delay += jitter
+
+            if max_delay is not None:
+                _delay = min(_delay, max_delay)
+
+
+def retry(
+    cancelation_token: threading.Event = threading.Event(),
+    exceptions: Iterable[Type[Exception]] = Exception,
+    tries: int = -1,
+    delay: int = 0,
+    max_delay: Optional[int] = None,
+    backoff: float = 1,
+    jitter: float = 0,
+    logger: logging.Logger = logging_logger,
+):
+    """Returns a retry decorator.
+
+    This is adapted from https://github.com/invl/retry
+
+    :param cancelation_token: a threading token that is waited on.
```

Comment:
We use google-style docstrings. Probably should have that in a contribution note somewhere

Reply:
I'll migrate them. This is straight up taken from the original repo.